docs/modules/ROOT/pages/upgrades.adoc

```diff
@@ -0,0 +1,113 @@
+
+
+[,javascript]
+----
+/// Upgrades the contract source code to the new contract class.
+fn _upgrade(new_class_hash: ClassHash) {
+    assert(!new_class_hash.is_zero(), 'Class hash cannot be zero');
+    starknet::replace_class_syscall(new_class_hash).unwrap();
+}
+----
+
+NOTE: If a contract is deployed without this mechanism, the source code is effectively immutable.
+
+== `Upgradeable` module
+
+OpenZeppelin Contracts for Cairo provides {upgradeable} to add upgradeability support to your contracts.
+
+=== Usage
+
+Integrating this module into a contract first requires supporting some access control logic, in order to
+avoid unauthorized upgrades. The {ownable} module is used for this example.
+
+[,javascript]
+----
+#[starknet::contract]
+mod UpgradeableContract {
+    use openzeppelin::access::ownable::Ownable;
+    use openzeppelin::upgrades::Upgradeable;
+    use starknet::ClassHash;
+    use starknet::ContractAddress;
+
+    #[storage]
+    struct Storage {}
+
+    #[constructor]
+    fn constructor(self: @ContractState, owner: ContractAddress) {
+        let mut unsafe_state = Ownable::unsafe_new_contract_state();
+        Ownable::InternalImpl::initializer(ref unsafe_state, owner);
+    }
+
+    #[external(v0)]
+    fn upgrade(ref self: ContractState, new_class_hash: ClassHash) {
+        // This function can only be called by the owner
+        let ownable_state = Ownable::unsafe_new_contract_state();
+        Ownable::InternalImpl::assert_only_owner(@ownable_state);
+
+        // Replace the class hash upgrading the contract
+        let mut upgradeable_state = Upgradeable::unsafe_new_contract_state();
+        Upgradeable::InternalImpl::upgrade(ref upgradeable_state, new_class_hash);
+    }
+
+    (...)
+}
+----
+
+=== Interface
+
+Implementing this interface to support upgradeability, supports interoperability by for example allowing DAOs
+and plugins to connect to contracts and upgrade them.
+
+[,javascript]
+----
+trait IUpgradeable {
+    fn upgrade(new_class_hash: ClassHash);
+}
+----
+
+== Proxies and Starknet
+
+A proxy contract is a contract that delegates function calls to another contract.
+This type of pattern decouples state and logic.
+Proxy contracts store the state and redirect function calls to an implementation contract that handles the logic,
+usually catching these calls with a fallback mechanism.
+This allows for different patterns such as upgrades, where implementation contracts can change but the proxy contract (and thus the state) does not;
+as well as deploying multiple proxy instances pointing to the same implementation.
+This can be useful to deploy many contracts with identical logic but unique initialization data.
+
+In the case of contract upgrades, it is achieved by simply changing the proxy's reference to the class hash of the declared implementation.
+This allows developers to add features, update logic, and fix bugs without touching the state or the contract address to interact with the application.
```

Comment:
who is this section for? my impression is that this is targeting people coming from the EVM (otherwise why explaining something that is not available in Starknet?). if that's the case, then i'd simplify these paragraphs assuming the reader knows about proxies already.

```suggestion
Proxies allow for different patterns such as upgrades as well as deploying multiple clones of the same contract. But Starknet achieves these two things in different ways.

In the case of contract upgrades, it is achieved by simply changing the contract's class hash of the declared implementation. in the case of clones, that's already how contracts and classes work.
```